### PR TITLE
Improved support for HOA

### PR DESCRIPTION
**Pull request by schafo**
HOA tags can now be extracted on parsing and support has been implemented for:
- Order
- Degree
- Normalization
- NfcRefDist
- ScreenRef

However, the `Equation` parameter remains unsupported.

### CODE REVIEW

File: include/adm/elements/screen_ref.hpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
+/// @file screen_ref.hpp
+#pragma once
+
+#include <boost/variant.hpp>
```
Comment (benjamin-weiss) [inline include/adm/elements/screen_ref.hpp]: I suppose this include and the export include can be removed?

File: src/private/rapidxml_formatter.cpp [hunk 2]
```diff
@@ -431,7 +431,12 @@ namespace adm {
       node.addAttribute<AudioBlockFormatId>(&audioBlock, "audioBlockFormatID");
       node.addOptionalAttribute<Rtime>(&audioBlock, "rtime");
       node.addOptionalAttribute<Duration>(&audioBlock, "duration");
-      // TODO: add missing hoa attributes and elements
+      node.addOptionalElement<Order>(&audioBlock, "order");
+      node.addOptionalElement<Degree>(&audioBlock, "degree");
+      node.addOptionalElement<NfcRefDist>(&audioBlock, "nfcRefDist");
+      node.addOptionalElement<ScreenRef>(&audioBlock, "screenRef");
+      // TODO: add Equation
+      // TODO: add Normalization
```
Comment (benjamin-weiss) [inline src/private/rapidxml_formatter.cpp]: Why is the normalisation added to the class but not written by the xml writer?

File: tests/xml_parser_audio_block_format_hoa_tests.h [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#ifndef XML_PARSER_AUDIO_BLOCK_FORMAT_HOA_TESTS_H
```
Comment (benjamin-weiss) [inline tests/xml_parser_audio_block_format_hoa_tests.h]: Why is this file necessary? Can this be removed?

File: src/private/rapidxml_formatter.cpp [hunk 4]
```diff
@@ -371,9 +371,9 @@ namespace adm {
       node.addOptionalElement<ChannelLock>(&audioBlock, "channelLock", &formatChannelLock);
       node.addOptionalElement<ObjectDivergence>(&audioBlock, "objectDivergence", &formatObjectDivergence);
       node.addOptionalElement<JumpPosition>(&audioBlock, "jumpPosition", &formatJumpPosition);
+      node.addOptionalElement<ScreenRef>(&audioBlock, "screenRef");
```
[the remark this answers is not in the document]
Comment (schafo) [reply src/private/rapidxml_formatter.cpp]: Aaah, thank you! I was a little confused as to why this was happening - I'll update that.

File: src/private/rapidxml_formatter.cpp (hunk 2 above)
Comment (schafo) [reply src/private/rapidxml_formatter.cpp]: This is just a miss by me. Normalization wasn't implemented at first, so I will just have forgotten to add it. Thanks for highlighting!

File: src/private/rapidxml_formatter.cpp [hunk 5]
```diff
@@ -431,7 +431,12 @@ namespace adm {
       node.addAttribute<AudioBlockFormatId>(&audioBlock, "audioBlockFormatID");
       node.addOptionalAttribute<Rtime>(&audioBlock, "rtime");
       node.addOptionalAttribute<Duration>(&audioBlock, "duration");
-      // TODO: add missing hoa attributes and elements
+      node.addOptionalElement<Order>(&audioBlock, "order");
+      node.addOptionalElement<Degree>(&audioBlock, "degree");
+      node.addOptionalElement<NfcRefDist>(&audioBlock, "nfcRefDist");
+      node.addOptionalElement<ScreenRef>(&audioBlock, "screenRef");
+      node.addOptionalElement<Normalization>(&audioBlock, "normalizaiton");
```
Comment (benjamin-weiss) [inline src/private/rapidxml_formatter.cpp]:
```suggestion
      node.addOptionalElement<Normalization>(&audioBlock, "normalization");
```

**Comment by schafo** [comment]
There's a bit more work to do and a few more changes, but just wanted to make sure we returned to a state with passing tests.

**Comment by schafo** [comment]
> I also wonder if it makes sense to update the AudioPackFormat class to support the HOA parameters too, while you are doing this. But I'm not sure yet what's the best way to add HOA support to the AudioPackFormat.

Sorry, only picking this up now and I've not been working on this for a while. Richard and I spoke about it a little - if it's just about adding basic support (e.g. adding HOA sub-elements as sub-elements in the existing `audio_pack_format`) I'm happy to do that. Anything else would probably be out of scope for what I'm doing at the moment unfortunately.

**Comment by benjamin-weiss** [comment]
I'm still a bit undecided. To have a consistent API we should have multiple `AudioPackFormats` as we have for the `AudioBlockFormats`, but that is really quite a bit of work.

I also wonder, if we should use a base class for these situations. Especially as the common part of the `AudioBlockFormat` got bigger in `-2`. It would probably be clearer and easier to maintain.

What do you think?

**Comment by schafo** [comment]
I agree on both points - for sake of consistency there should be really be multiple `AudioPackFormat`. A base class seems like a good approach, especially in this case where it's only applicable to HOA and Matrix.

The thing which might start to add complexity is the checks for if sub-elements of `audioPackFormat` and `audioBlockFormat` match - and having the `audioBlockFormat` take priority. This is part of what I feel is out of scope at the moment.
